flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -98,6 +104,15 @@ public class TaskLocalStateStoreImpl implements OwnedTaskLocalStateStore {
     @GuardedBy("lock")
     private final SortedMap<Long, TaskStateSnapshot> storedTaskStateByCheckpointID;
 
+    /** The relationship between checkpoints. (cp1,cp3) means cp3 refer to cp1's state. */
+    private final Map<Long, Long> referredByCheckpointID;
+
+    /**
+     * (MaterializationID, created by checkpointID). Because local store only keeps one checkpoint,
+     * at most one materialization can exist at the same time.
+     */
+    private Tuple2<Long, Long> currentMaterializationID;
+
     public TaskLocalStateStoreImpl(
```

Comment:
Maybe considering to extract some codes related to changelog to a new subclass?

Reply:
Nice suggestion, I created ChangelogTaskLocalStateStore to encapsulate changelog-specific logic.